shared/src/main/scala/mlscript/ConstraintSolver.scala

```diff
@@ -698,5 +740,107 @@ class ConstraintSolver extends NormalForms { self: Typer =>
     freshen(ty)
   }
   
-  
+  // a is indexed by i
+  // i is indexed in a
+  // receiver: a, index: i, return type of a[i]
+  def constrainIndex(receiver: SimpleType, index: SimpleType)(implicit ctx: Ctx, raise: Raise): SimpleType 
+        = trace(s"$lvl. Receiver: $receiver, Index: $index") {
+    (receiver.unwrapProxies, index.unwrapProxies) match {
+      // if one of receiver, index is Error, then return error
+      case (_, e @ ClassTag(ErrTypeId, _)) => e
+      case (e @ ClassTag(ErrTypeId, _), _) => e
+      // TupleType: fixed length of array, inherited from ArrayBase
+      case (t @ TupleType(fs), ClassTag(IntLit(value), _)) =>  
+        // check index validity and retrieve corresponding type
+        if (value >= fs.length || value < 0) {
+          err(msg"Out of range!", t.prov.loco)
+        } else {
+          fs(value.toInt)._2.ub
+        }
+      case (t: ArrayBase, ClassTag(Var("int"), _)) =>  
+        t.inner.ub | TypeRef(TypeName("undefined"), Nil)(noProv)
+      case (t: ArrayBase, ClassTag(IntLit(value), _) ) =>
+        if (value < 0) {
+          err(msg"Out of range!", t.prov.loco)
+        } else {
+          t.inner.ub | TypeRef(TypeName("undefined"), Nil)(noProv)
+        }
+      // for the case with type ascription, like k: T, it can be assigned a value or not
```

Comment:
I don't understand what this comment is trying to say. Remove?

Reply:
like we can `a: int` (without being assigned a value) or `a = 3: int` (being assigned a value but its type becomes int instead of 3, which hides information about the underlying value)

Reply:
In other words, whether it's "assigned a value" is totally irrelevant and I don't know why you're fixating on it. All that matters is whether the static type is precise or not.